src/llama-impl.cpp

```diff
@@ -84,14 +84,16 @@ std::string format(const char * fmt, ...) {
     va_list ap2;
     va_start(ap, fmt);
     va_copy(ap2, ap);
-    int size = vsnprintf(NULL, 0, fmt, ap);
+    const int size = vsnprintf(NULL, 0, fmt, ap);
     GGML_ASSERT(size >= 0 && size < INT_MAX); // NOLINT
-    std::vector<char> buf(size + 1);
-    int size2 = vsnprintf(buf.data(), size + 1, fmt, ap2);
+    std::string buf;
+    buf.resize(size);
+    const int size2 = vsnprintf(const_cast<char *>(buf.data()), buf.size() + 1, fmt, ap2);
```

Comment:
Isn't this UB?

https://en.cppreference.com/w/cpp/string/basic_string/data

Reply:
As long as we don't modify past the null terminator (or .size())in the std::string, I think it's fine. At least I've done this kinda thing in various codebases for 10 years and it's never been an issue.

I think:

```
&buf[0]
```

is also an option.

But if we don't want to do this it's fine, it just seems sub-optimal to create a vector<char> (which is essentially a poor mans std::string) and then copy all the data to a std::string.

Reply:
The `+ 1` is fine because the std containers do not explicitly need room for the null terminating character. This is in contrast to plain C string `char buf[size]` where the string length can be maximum `size - 1`.

Regarding the UB, I'm also sure that this will always work in the real world, but would still prefer to avoid adding UB to the code (in case this is indeed UB, not sure yet). In this case the performance consideration is not very important, because we use `fmt` to just print error messages and logs, so it's OK to make the copy.

Reply:
I removed the micro-optimization in case of UB. Makes the review easier for everyone.